core/src/utils/mod.rs

```diff
@@ -19,6 +19,352 @@ pub mod math {
+                        }
+                        return Ok(output);
+                    }
+                }
+            }
+        }
+
+        // Check if we should reset our internal buffer
+        if self.buffer().is_empty() && self.pos > 0 {
+            unsafe {
+                self.buf.set_len(0);
+            }
+        }
+
+        Ok(output)
+    }
+
+    /// Fill `self.buf` with `count` bytes
+    ///
+    /// This should only be called when we can't read from the reader directly
+    fn buffer_at_least(&mut self, mut count: usize) -> Result<(), DeserializationError> {
+        loop {
+            if count == 0 || self.buf.len() >= count {
+                break Ok(());
+            }
+            self.non_empty_reader_buffer_mut()?;
+            // We have to re-borrow the reader buffer here
+            // to copy bytes between the two buffers.
+            let reader = self.reader.get_mut();
+            let buf = reader.buffer();
+            let consumed = buf.len();
+            self.buf.extend_from_slice(buf);
+            reader.consume(consumed);
+            count = count.saturating_sub(consumed);
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<'a> ByteReader for ReadAdapter<'a> {
+    #[inline]
+    fn read_u8(&mut self) -> Result<u8, DeserializationError> {
+        self.pop()
+    }
+    /// NOTE: If we happen to not have any bytes buffered yet
+    /// when this is called
+    fn peek_u8(&self) -> Result<u8, DeserializationError> {
+        if let Some(byte) = self.buffer().first() {
+            return Ok(*byte);
+        }
+        self.non_empty_reader_buffer().map(|b| b[0])
+    }
+    fn read_slice(&mut self, len: usize) -> Result<&[u8], DeserializationError> {
+        // Edge case
+        if len == 0 {
+            return Ok(&[]);
+        }
+
+        // If we have unused buffer, and the consumed portion is
+        // large enough, we will move the unused portion of the buffer
+        // to the start, freeing up bytes at the end for more reads
+        // before forcing a reallocation
+        let should_optimize_storage = self.pos >= 16 && !self.has_remaining_capacity(len);
+        if should_optimize_storage {
+            // We're going to optimize storage first
+            let buf = self.buffer();
+            let src = buf.as_ptr();
+            let count = buf.len();
+            let dst = self.buf.as_mut_ptr();
+            unsafe {
+                core::ptr::copy(src, dst, count);
+                self.buf.set_len(count);
+                self.pos = 0;
+            }
+        }
+
+        // Fill the buffer so we have at least `len` bytes available,
+        // this will return an error if we hit EOF first
+        self.buffer_at_least(len)?;
+
+        Ok(&self.buffer()[0..len])
+    }
+    #[inline]
+    fn read_array<const N: usize>(&mut self) -> Result<[u8; N], DeserializationError> {
+        if N == 0 {
+            return Ok([0; N]);
+        }
+        self.read_exact()
```

Comment:
The value of `N` is inferred because `read_exact` return type is unified with `read_array` return type. This happens because the value returned by `read_exact` is used as the result of `read_array`, and their types must match.

```rust
fn read_array<const N: usize>(&mut self) -> Result<[u8; N], DeserializationError>
fn read_exact<const N: usize>(&mut self) -> Result<[u8; N], DeserializationError>
```

Reply:
Interesting! Thank you for the explanation! 
Before I wrote this comment I tried to implement this in the playground and it returned and error, but I suppose that playground just uses an old version of rust.

Reply:
Depending on how you defined the example, it might've complained due to not knowing the specific integer type in question, rather than not being able to infer the dimension of the array, but it's hard to say. Various things can interfere with inference, but in simple cases like this bit of code, it's unambiguous.